skybar/src/main/java/org/wtf/skybar/registry/SkybarRegistry.java

```diff
@@ -23,106 +23,77 @@
-        } finally {
-            phaser.writerCriticalSectionExit(l);
+    public void registerLine(String sourceName, int lineNumber) {
+        Map<Integer, LongAdder> lines = visits.get(sourceName);
+        if(lines == null) {
+            Map<Integer, LongAdder> newLines = new HashMap<>();
+            Map<Integer, LongAdder> existingLines = visits.putIfAbsent(sourceName, newLines);
+            lines = existingLines != null ? existingLines : newLines;
         }
-
-        return index;
+        lines.putIfAbsent(lineNumber, new LongAdder());
     }
 
     /**
-     * @param deltaBuffer map to use to write delta into
+     * Can only be called after registerLine has been called for the same line
+     * @param sourceName name of the source file
+     * @param lineNumber the line number to get the LongAdder for
+     * @return the LongAdder used to accumulate visit counts for the line
      */
-    public void updateListeners(Map<String, IntLongMap> deltaBuffer) {
-        deltaBuffer.forEach((s, counts) -> counts.clear());
-
-        synchronized (this) {
-            flipPhase();
-
-            inactiveCounts.forEach((index, adder) -> {
-
-            /*
-             this is accessed outside the registry's phaser read lock, but all keys we will access here for the
-             delta must already have been added at registration time
-            */
-                SourceLocation location = indexToSourceLoc.get(index);
+    public LongAdder getAdderForLine(String sourceName, int lineNumber) {
+        return visits.get(sourceName).get(lineNumber);
+    }
 
-                BiFunction<String, IntLongMap, IntLongMap> mapUpdater
-                        = getMapUpdater(location.lineNum, adder);
-                deltaBuffer.compute(location.source, mapUpdater);
-                accumulatedCounts.compute(location.source, mapUpdater);
-            });
+    /**
+     * @param prev map containing the previously accumulated line visits
+     */
+    public void updateListeners(Map<String, IntLongMap> prev) {
+        Map<String, IntLongMap> delta = new HashMap<>();
+        for (Map.Entry<String, Map<Integer, LongAdder>> sourceLines : visits.entrySet()) {
+
+            String sourceName = sourceLines.getKey();
+            Map<Integer, LongAdder> currentLines = sourceLines.getValue();
+
+            Map<Integer, Long> prevLines = prev.get(sourceName);
+
+
+            if(prevLines == null) {
+                IntLongMap map = HashIntLongMaps.newMutableMap();
+                delta.put(sourceName, map);
+                prev.put(sourceName, map);
+                currentLines.forEach((lnum, adder) -> map.put(lnum.intValue(), adder.longValue()));
+            } else {
+                currentLines.forEach((lnum, adder) -> {
+                    long count = adder.longValue();
+                    Long prevCount = prevLines.get(lnum);
+                    long diff = count - (prevCount == null ? 0 : prevCount);
```

Comment:
Can build a IntLongMap that has a 0 default value for missing keys; then we won't need a null check (or a conditional at all -- can just subtract prevCount and make the branch predictor happy).